crates/ruma-client-api/src/knock/knock_room.rs

```diff
@@ -23,22 +23,130 @@ pub mod v3 {
     };
 
     /// Request type for the `knock_room` endpoint.
-    #[request(error = crate::Error)]
+    #[derive(Clone, Debug)]
+    #[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
     pub struct Request {
         /// The room the user should knock on.
-        #[ruma_api(path)]
         pub room_id_or_alias: OwnedRoomOrAliasId,
 
         /// The reason for joining a room.
-        #[serde(skip_serializing_if = "Option::is_none")]
         pub reason: Option<String>,
 
         /// The servers to attempt to knock on the room through.
         ///
         /// One of the servers must be participating in the room.
-        #[ruma_api(query)]
+        pub via: Vec<OwnedServerName>,
```

Comment:
We add in the docs that it serializes both as `server_name` and `via`, and prefers `via` during deserialization, so users don't have to look at the code to see what is happening.